src/platform/OpenThread/GenericThreadBorderRouterDelegate.cpp

```diff
@@ -0,0 +1,216 @@
+
+class ScopedThreadLock
+{
+public:
+    ScopedThreadLock() { DeviceLayer::ThreadStackMgr().LockThreadStack(); }
+    ~ScopedThreadLock() { DeviceLayer::ThreadStackMgr().UnlockThreadStack(); }
+};
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::Init()
+{
+    mCallback = nullptr;
+    ReturnErrorOnFailure(DeviceLayer::PlatformMgrImpl().AddEventHandler(OnPlatformEventHandler, reinterpret_cast<intptr_t>(this)));
+    ReturnErrorOnFailure(RevertActiveDataset());
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetBorderAgentId(MutableByteSpan & borderAgentIdSpan)
+{
+    otInstance * otInst = DeviceLayer::ThreadStackMgrImpl().OTInstance();
+    VerifyOrReturnError(otInst, CHIP_ERROR_INCORRECT_STATE);
+    otBorderAgentId borderAgentId;
+    if (borderAgentIdSpan.size() < sizeof(borderAgentId.mId))
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    ScopedThreadLock threadLock;
+    otError err = otBorderAgentGetId(otInst, &borderAgentId);
+    if (err == OT_ERROR_NONE)
+    {
+        memcpy(borderAgentIdSpan.data(), borderAgentId.mId, sizeof(borderAgentId.mId));
+        borderAgentIdSpan.reduce_size(sizeof(borderAgentId.mId));
+        return CHIP_NO_ERROR;
+    }
+    return CHIP_ERROR_INTERNAL;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetThreadVersion(uint16_t & threadVersion)
+{
+    threadVersion = otThreadGetVersion();
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetInterfaceEnabled(bool & interfaceEnabled)
+{
+    otInstance * otInst = DeviceLayer::ThreadStackMgrImpl().OTInstance();
+    VerifyOrReturnError(otInst, CHIP_ERROR_INCORRECT_STATE);
+    ScopedThreadLock threadLock;
+    interfaceEnabled = otIp6IsEnabled(otInst);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetDataset(Thread::OperationalDataset & dataset, DatasetType type)
+{
+    otInstance * otInst = DeviceLayer::ThreadStackMgrImpl().OTInstance();
+    VerifyOrReturnError(otInst, CHIP_ERROR_INCORRECT_STATE);
+
+    ScopedThreadLock threadLock;
+    otError otErr = OT_ERROR_NONE;
+    otOperationalDatasetTlvs datasetTlvs;
+    if (type == DatasetType::kActive)
+    {
+        otErr = otDatasetGetActiveTlvs(otInst, &datasetTlvs);
+    }
+    else
+    {
+        otErr = otDatasetGetPendingTlvs(otInst, &datasetTlvs);
+    }
+    if (otErr == OT_ERROR_NONE)
+    {
+        return dataset.Init(ByteSpan(datasetTlvs.mTlvs, datasetTlvs.mLength));
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::SetActiveDataset(const Thread::OperationalDataset & activeDataset,
+                                                                   uint32_t sequenceNum, ActivateDatasetCallback * callback)
+{
+    CHIP_ERROR err = BackupActiveDataset();
+    if (err == CHIP_NO_ERROR)
+    {
+        err = DeviceLayer::ThreadStackMgrImpl().AttachToThreadNetwork(activeDataset, nullptr);
```

Comment:
The current specification seems to suggest that after the Active dataset is configured the BR will join/form the Thread network, but I am wondering if it's useful to have some sort of control over this operation? Maybe the application running the TBR would want to control when to form/join the Thread network. 
Thoughts?

Reply:
> The current specification seems to suggest that after the Active dataset is configured the BR will join/form the Thread network.

Yes, The SPEC says that `If this command is invoked and the ActiveDatasetTimestamp attribute is null, the Thread Border Router SHALL configure and activate its active dataset using the ActiveDataset parameter.`
 
> if it's useful to have some sort of control over this operation.

Agree, but the TBR can still implement another delegate to control when to form/join the Thread network. Currently the GenericDelegate is enough for our application.

Reply:
Let's follow-up

Reply:
#34596